### PR TITLE
autograd: add executor to cache key for make_aug_forward_and_backward

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes: https://github.com/Lightning-AI/lightning-thunder/issues/230

### CODE REVIEW

File: thunder/tests/test_grad.py [hunk 1]
```diff
@@ -1464,3 +1464,41 @@ def func(a, b):
         fw_out = cfunc(a, b)
 
     thunder.extend.deregister_executor(myex)
+
+
+def test_make_forward_backward_symbol_caching_with_executor():
```
Comment (IvanYashchuk) [inline thunder/tests/test_grad.py:1469]: Nice test!

File: thunder/core/vjp_utils.py [hunk 2]
```diff
@@ -47,15 +47,14 @@ def make_aug_forward_and_backward(bsym: BoundSymbol) -> tuple[Callable, Callable
...
 
-    joint_forward_backward = _get_gradfn(bsym)
+    joint_forward_backward, executor = _get_gradfn_and_executor(bsym)
     utils.check(
         joint_forward_backward is not None,
         lambda: f"Cannot generate forward and backward functions for {bsym.sym.name}",
     )
-
-    key = (bsym.sym, subkey := _make_cache_key(bsym.args, bsym.kwargs))
+    key = (bsym.sym, subkey := _make_cache_key(bsym.args, bsym.kwargs), executor)
```
Comment (IvanYashchuk) [inline thunder/core/vjp_utils.py:57]: Here's a small readability improvement and potential optimization in the case of a cache miss. I don't know the details but I think to determine a cache miss faster it's better to put elements that have a higher chance to change as the first ones.
```suggestion
    key = (bsym.sym, executor, subkey := _make_cache_key(bsym.args, bsym.kwargs))
```

**Review by IvanYashchuk — APPROVED** [review]
Looks good! We should merge it.

File: thunder/core/transforms.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 from functools import lru_cache, partial, wraps
 import math
 from numbers import Number
-from typing import Any, Dict, Union, Optional
+from typing import Any, Dict, Union, Optional, Tuple
```
Comment (crcrpar) [inline thunder/core/transforms.py:9]:
```suggestion
from typing import Any, Dict, Union, Optional
```

**Review by crcrpar — APPROVED** [review]

**Comment by kshitij12345** [comment]
@t-vi can you please have a look and stamp this. Thanks!

**Review by carmocca — APPROVED** [review]

**Pull request merged**